### ISSUE TITLE
istio-ctl proxy-status <envoy> constantly shows diffs

### ISSUE THREAD

**Issue by Evesy** [istio/istio#10293]
**Describe the bug**
`istioctl proxy-status` shows Envoy instance as fully synced:

```
$ istioctl proxy-status
PROXY                                      CDS        LDS        EDS               RDS        PILOT                            VERSION

monkeynetes-69f85c5749-rbw9c.monkeynetes   SYNCED     SYNCED     SYNCED (100%)     SYNCED     istio-pilot-5c8c66d8ff-r2m4n     1.0.2
```

`istioctl proxy-status monkeynetes-69f85c5749-rbw9c.monkeynetes` constantly displays diffs (for a hanful of different routes each time) between itself and Pilot. The diff as shown below is actually the same config however ordering of items in `requestHeadersToAdd` is different:

```
                         "route": {
                            "cluster": "outbound|80||app.e2e-381-deploy-dont-purge-existing-rels.svc.cluster.local",
                            "timeout": "10.000s",
                            "requestHeadersToAdd": [
                               {
                                  "header": {
+                                    "key": "x-envoy-max-retries",
+                                    "value": "10"
+                                 }
+                              },
+                              {
+                                 "header": {
                                     "key": "x-envoy-retry-on",
                                     "value": "gateway-error,connect-failure,refused-stream"
-                                 }
-                              },
-                              {
-                                 "header": {
-                                    "key": "x-envoy-max-retries",
-                                    "value": "10"
                                  }
                               }
                            ],
                            "maxGrpcTimeout": "10.000s"
                         },
```

**Expected behavior**
Routes should be identified as matching.

**Steps to reproduce the bug**
- Pilot with PILOT_DIRECT_EDS enabled
- Have Istio enabled pod
- Have another service w/ a matching VirtualService:
  ```
  apiVersion: networking.istio.io/v1alpha3
  kind: VirtualService
  metadata:
    name: app
    namespace: e2e-381-deploy-dont-purge-existing-rels
  spec:
    hosts:
    - app
    http:
    - appendHeaders:
        x-envoy-max-retries: "10"
        x-envoy-retry-on: gateway-error,connect-failure,refused-stream
      route:
      - destination:
          host: app
        weight: 100
      timeout: 10s
  ```
- `istioctl proxy-status <pod>` incorrectly shows diffs
- Possibly related? https://github.com/istio/istio/pull/10177

**Version**
Istio:
```
Version: 1.0.4
GitRevision: d5cb99f479ad9da88eebb8bb3637b17c323bc50b
User: root@8c2feba0b568
Hub: docker.io/istio
GolangVersion: go1.10.4
BuildStatus: Clean
```

Kubernetes:
```
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.7-gke.11", GitCommit:"fa90543563c9cfafca69128ce8cd9ecd5941940f", GitTreeState:"clean", BuildDate:"2018-11-08T20:22:21Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Installation**
via Helm, context: #3720

**Environment**
GKE, cos

**Cluster state**
Can provide if needed.

**Comment by linsun** [comment]
cc @ericvn

**Comment by ericvn** [comment]
@Evesy Hello, I am trying to replicate this issue and trying to better understand the scenario.

The `istioctl` output shows version 1.0.2 while version 1.0.4 was mentioned. Did you possibly upgrade from 1.0.2 to 1.0.4?

I believe in your steps you have two Istio enabled services (steps two and three) and one of them (step three) has a virtual service. Is that correct?

Thanks.

**Comment by Evesy** [comment]
@ericvn Sorry, the output I wasn't too clear. `1.0.4` is the version of `istioctl`.

Both Service A and Service B are Istio enabled services, both of which have `VirtualService`'s and `DestinationRule`'s.

When running `istioctl proxy-status` against a specific pod/envoy instance, it shows constant diffs in the Envoy config on (from what I've observed) all routes that have a non-empty `requestHeadersToAdd` array:

```
                         "route": {
                            "cluster": "outbound|9080||app.tdt.svc.cluster.local",
                            "timeout": "10.000s",
                            "requestHeadersToAdd": [
                               {
                                  "header": {
-                                    "key": "x-envoy-retry-on",
-                                    "value": "gateway-error,connect-failure,refused-stream"
-                                 }
-                              },
-                              {
-                                 "header": {
-                                    "key": "x-envoy-max-retries",
-                                    "value": "10"
+                                    "key": "x-envoy-max-retries",
+                                    "value": "10"
+                                 }
+                              },
+                              {
+                                 "header": {
+                                    "key": "x-envoy-retry-on",
+                                    "value": "gateway-error,connect-failure,refused-stream"
                                  }
                               }
                            ],
                            "maxGrpcTimeout": "10.000s"
                         },
```

**Comment by ericvn** [comment]
Thanks @Evesy. I was able to reproduce your issue on 1.0.4 using the BookInfo sample and tweaking the _reviews_ virtual service resulting in intermittent failures when doing a `istioctl proxy-status` against _ratings_.

I am trying the same scenario with the current 1.1 builds and have been unsuccessful so far. I need to investigate more but wondering if https://github.com/istio/istio/pull/10233 might have resolved your issue in the 1.1 stream.

**Comment by Evesy** [comment]
Great, thanks @ericvn, that certainly looks better!

Does this mean in this setup with 1.0.4 Envoy is very frequently reloading to pull in these 'changes' or is it purely just a visual manifestation when running `proxy-status`?

**Comment by emedina** [comment]
Not sure whether this fits here, but I'm defining the following `VirtualService` in `1.1-rc.0` for `httpbin`:

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: httpbin
  namespace: default
spec:
  gateways:
  - httpbin-gateway
  hosts:
  - httpbin.org
  http:
   - appendHeaders:
      x-envoy-max-retries: "10"
      x-envoy-retry-on: gateway-error,connect-failure,refused-stream
    match:
    - uri:
        prefix: /
    retries:
      attempts: 5
      retryOn: 5xx
    route:
    - destination:
        host: httpbin
        port:
          number: 8000
```

But when I check the `config_dump` endpoint for the `istio-proxy` deployed with `httpbin`, I cannot find any of either the headers or the retries in such config:

```
# kubectl exec -ti httpbin-54f5bb4957-m7v7r -c istio-proxy -- curl localhost:15000/config_dump

    {
     "version_info": "2019-02-27T09:34:40Z/91",
     "cluster": {
      "name": "inbound|8000||httpbin.default.svc.cluster.local",
      "connect_timeout": "1s",
      "circuit_breakers": {
       "thresholds": [
        {}
       ]
      },
      "load_assignment": {
       "cluster_name": "inbound|8000||httpbin.default.svc.cluster.local",
       "endpoints": [
        {
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "127.0.0.1",
              "port_value": 80
             }
            }
           }
          }
         ]
        }
       ]
      }
     },
     "last_updated": "2019-02-27T09:35:43.959Z"
    },
```

But maybe I'm missing something here.

**Comment by esnible** [comment]
@emedina your problem is not related to this issue.

You should check the config_dump not of httpbin (the target), but of the istio-ingressgateway pod in the istio-system namespace.  It is that pod which appends the headers as it invokes httpbin.  The implementation of appendHeaders is to have them done by the outbound proxy, rather than the inbound proxy.

**Comment by emedina** [comment]
@esnible Thanks for the advice. And sorry about the noise.

**Comment by linsun** [comment]
is this still an issue for 1.1 @ericvn @Evesy ?

**Comment by ericvn** [comment]
I tested again using the master branch my scenario above. I also tested it with more headers (more headers should yield more permutations which would differ causing a diff to appear). I did not see any diffs displayed. I did look through the code to see if I could find anywhere changes where made so the strings would be serialized in a sorted order, instead of random (the problem @esnible mentioned above). I didn't see a specific change although a change could have been made in the libraries we vendor.

For now, I think the problem is resolved (and was probably resolved in 1.1 as well) and will close this issue. It can be re-opened if the problem re-appears.
